Introduce Analytics API, core events, and a GA implementation provider

iamEAP opened the pull request:
## What

- **Introduces an Analytics API** in `core-plugin-api` with 4 key concepts:
  - The API is built on events. Events consist of, at a minimum, an `action` (e.g. `click`) and a `subject` (e.g. `thing-that-was-clicked-on`)
  - Plugin Devs can capture events using a tracker returned by the hook `useAnalytics()`.
  - Event capture pulls in metadata from the react tree above it, defined using `<AnalyticsContext />`, whose attributes inherit from any context above.  Three "well-known" attributes are provided automatically: `pluginId`, `routeRef`, and `extension`.
  - The API is optional and does nothing with captured events until an App Integrator provides a concrete implementation.
- **Provides Extension-Based Analytics Contexts**, including key metadata like Plugin IDs and Extension Names.
- **Instruments `core-app-api`** with a `navigate` event
- **Instruments `<Link />` component** in `core-components` with a `click` event
- **Adds `@backstage/plugin-analytics-provider-ga` plugin** which can be used to send instrumented events to Google Analytics in an idiomatic way.

## Why

- **Why this API at all**: See the need in the RFC (#6432)
- **Why navigate/click events**: Based on experience internally, having a foundational layer of "what are people looking at" and "what are people clicking on" _that can be segmented by plugin_ is, in 80% of cases, good enough to help understand and prioritize investments in different areas of a Backstage instance. Building these events on a common framework that can be used by Plugin Devs to instrument more specific things enables the remaining 20% of cases.
- **Why GA Analytics Provider**: Beyond just being useful out the gate, the provider demonstrates how events, event context, and analytics domains can be combined in a way that is useful within the confines of a fairly constrained analytics system (compared with others where simply forwarding all events as JSON objects to an endpoint might be all that's needed).

## Follow-Ups
- [ ] Add `name` keys to all extensions in the monorepo (matching their export name)
- [ ] Create an issue to clean up analytics API in core plugin API (manage context via version bridge + account for routable extenions with data manually attached)
- [ ] Create an issue to clean up tracker implementation in core app API (don't duplicate tree traversal, access from app provider instead)
- [ ] Create an issue to extend the analytics API so that user-level information can also be captured/tracked
- [ ] Create issues (label help wanted) for instrumenting some of the primary actions in the core plugins.
- [ ] Document the Analytics API (thoroughly)
- [ ] Add the GA Analytics provider to the plugin marketplace

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

freben commented on .changeset/analytics-glimmering-radio-station.md:
```diff
@@ -0,0 +1,9 @@
+---
+'@backstage/plugin-analytics-provider-ga': patch
```
```suggestion
'@backstage/plugin-analytics-module-ga': patch
```

The naming does of course make sense, but the `module` qualifier is starting to get a foothold

dhenneke commented on plugins/analytics-module-ga/README.md:
```diff
@@ -0,0 +1,131 @@
+# app-config.yaml
+app:
+  analytics:
+    provider: ga
```
What is the intent behind this `provider` config property? When one adds the `@backstage/plugin-analytics-module-ga` dependency and setups the plugin export, it is always used regardless of this property.

iamEAP replied:
I suppose I was thinking to follow a similar pattern in other Backstage plugins where there's a key that allows one to differentiate config schemas under a common key.  Like...

```typescript
export interface Config = {
  app: {
    analytics?:
      // Google Analytics
      | {
        provider: 'ga',
        ga: {
          trackingId: string;
        }
      // Amplitude
      | {
        provider: 'amplitude';
        amplitude: {
          apiKey: string;
        };
      };
  };
};
```

But maybe this isn't necessary or valuable if the configs are spread across different plugins.

iamEAP replied:
Ended up removing this.

dhenneke commented on plugins/analytics-module-ga/README.md:
```diff
@@ -0,0 +1,131 @@
+
+## Installation
+
+1. Install this plugin in your Backstage instance: `yarn add @backstage/plugin-analytics-module-ga`
```
You could add that this should be installed at the `/packages/app/package.json` level and not the root level `/package.json`.

iamEAP replied:
Good call.

freben commented:
Gentle ping here too. Where are we on this one?

iamEAP commented:
> Gentle ping here too. Where are we on this one?

Would love your input/review!

iamEAP commented:
(Rebased, and simplified the API a bit.  _Now_ I'd love a look / feedback @freben 😬 )

venkiit58 commented:
When can we expect this feature to be available? Please.

joaomilho commented:
Did you take a look at https://github.com/segmentio/analytics.js/blob/master/lib/integrations.js ? They already built an abstraction for this long ago.

freben commented:
That does look like a pretty good abstraction layer.

freben commented:
Hi, I just wanted to say that it sounds good to not worry about analytics.js, and that we want to do a mob review among ourselves of this so we get it through as efficiently as possible, and will get back to ya.

iamEAP commented on packages/core-app-api/src/routing/RouteTracker.tsx:
```diff
@@ -0,0 +1,121 @@
+ * If no exact match is found, path parts are stripped away, one-by-one, until
+ * a parent-level path matches a route.
+ */
+const getExtensionDomain = (
```
There must be an easier way to do this that I'm not aware of...?

iamEAP replied:
> _It could be something to explore, as in adding this to the routable extensions instead._

Would the idea be that tracking be performed on render of a routable extension?  I guess in that case, the disadvantage would be that any routes added to the router that are not routable extensions wouldn't be captured?

> _This is how it's used by the RouteResolver to look up a common base RouteRef during routing_

I may need to sit with this one awhile to understand fully.  It's almost like what I want is a Backstage equivalent of react router's `createRoutesFromChildren()`, which returns `BackstageRouteObjects[] | null`, and which I can just run a simple `matchRoutes()` on with a given path.  

Am I on the right track, here?

Rugvip replied:
> Would the idea be that tracking be performed on render of a routable extension? I guess in that case, the disadvantage would be that any routes added to the router that are not routable extensions wouldn't be captured?

Yep, and yeah it is a disadvantage, it does mean you'll need to stick to the correct type of extension within the route elements if you want tracking.

> Am I on the right track, here?

Yep! We do a big traversal of the elements at startup with a bunch of "collectors". They're responsible for collecting different pieces of information for usage later on. One of them is the [`routeObjectCollector`](https://github.com/backstage/backstage/blob/802f3f106b937794e2da7eeb74086481618bd3d2/packages/core-app-api/src/routing/collectors.tsx#L124), which is essentially `createRoutesFromChildren()`, but the objects also include RouteRefs and could include more stuff like plugins.

iamEAP replied:
Think I figured out a reasonable way to do this!  Thanks for the pointers.  See updates

iamEAP commented on packages/core-plugin-api/src/extensions/extensions.tsx:
```diff
@@ -37,6 +38,7 @@ export function createRoutableExtension<
 >(options: {
   component: () => Promise<T>;
   mountPoint: RouteRef;
+  name?: string;
```
It turns out that the lazy-loading logic in here means that when `createReactExtension()` is ultimately called, the underlying component never gets properly named (because `displayName` and `name` are never set on a lazy-loaded object).  ...All of these changes were my way of giving plugin developers a way to explicitly set a component name at extension registration-time.

Rugvip replied:
To be honest we're a bit unsure about the value of the component name in the analytics. It seems something that is a bit implementation specific and also might change over time in unexpected ways. How do you feel about leaving that piece out for new? :grin:

iamEAP replied:
Yeah, lemme talk through motivations real quick, and then I'd love to hear your thoughts:

Internally, I've found it very handy to be able to slice and dice events at a variety of levels of detail.  

Sometimes you just want to take a really shallow look at aggregate events of all kinds happening under a plugin.  For example, a PM might ask, _"What's the Catalog MAU?"_ -- and one could answer with... `SELECT DISTINCT COUNT(userId) WHERE pluginId='catalog' AND date >= [30 Days Ago]`.

Other times, you need to be able to distinguish between two otherwise identical events (e.g. same verb/action + noun/subject) where the only difference is the immediate context of the interaction on the page.  For example, a PM might ask, _"Do people on Catalog Entity pages click to the entity's owner more by clicking the link in the About card, or in the header?"_  -- which you might be able to answer with...  `SELECT componentName, COUNT(*) WHERE verb='click' AND noun LIKE '/catalog/default/group/%' AND routeRef='Entity' GROUP BY componentName`.

When the App Integrator asking the questions and the Plugin Developer implementing the instrumentation are one in the same, you have the luxury of being able to just add extra contextual metadata to the events to be able to distinguish them, but then here we are with Backstage!  So having those levels of detail (what plugin? what page? what part of the page?) instrumented out-of-the-box is handy.

My hope was to have a system such that, just by building your plugin "the right way" (making use of extensions / the composability API / etc.), you would get sufficient levels of detail to handle most such cases.  ...Only after playing around with it, did I realize that component name was kind of broken (basically everything is just `Component`).

Any other ideas on how to achieve the design goal above?

Rugvip replied:
What I'm thinking is that it's the "extension name" that we want rather than the component name, which tbh is pretty much exactly this just that I thought about it from the other direction 😅

It's a lil bit unfortunate if we have to repeat the name for every extension declaration. I'm thinking we might be able to just grab the display name of the extension component itself as we traverse the element tree. Prolly works well with the plugin and route gathering as well?

iamEAP replied:
Alright, I think by just renaming `componentName` to `extension`, we get where we want.  See top-level comment (forthcoming) on this PR for full details.

iamEAP commented:
> we want to do a mob review among ourselves of this so we get it through as efficiently as possible, and will get back to ya. ~@freben

Cool!  I went through, rebased, cleaned up a few things I found, added notes about weird-looking changes, etc.  Lemme know if it'd be helpful to break this apart into different PRs.  I could strip out the GA stuff, for example.

iamEAP commented on packages/core-app-api/src/routing/RouteTracker.tsx:
```diff
@@ -0,0 +1,121 @@
+/**
+ * Performs the actual event capture on render.
+ */
+const CaptureOnRender = ({
```
Perhaps just `<TrackNavigation />`

Rugvip commented on packages/core-plugin-api/src/analytics/useAnalytics.tsx:
```diff
@@ -0,0 +1,58 @@
+function useTracker(): AnalyticsTracker {
+  const analyticsApi = useApi(analyticsApiRef);
+  const domain = useAnalyticsDomain();
+  return {
```
Should `useMemo` this to make sure that we don't need to return a new instance too often. Another option that might be nicer is to create a small class that does this wrapping and then store an instance of that here with a `useRef`. You'd then have some method on the class that we can use to mutate in as the `domain` changes, keeping the returned `AnalyticsTracker` reference completely stable.

Rugvip commented on packages/core-plugin-api/src/analytics/useAnalytics.tsx:
```diff
@@ -0,0 +1,58 @@
+          domain,
+        });
+      } catch {
+        // Just don't throw on an instrumentation problem.
```
Would argue that we should log at least as a warning tho

Rugvip commented on packages/core-plugin-api/src/analytics/useAnalytics.tsx:
```diff
@@ -0,0 +1,58 @@
+export function useAnalytics(): AnalyticsTracker {
+  // Return a no-op tracker if no implementation for the Analytics API is
+  // available. Having no default Analytics API implementation enables simple
+  // provider installation via plugin instantiation.
```
Thinking that we should just have a default no-op `AnalyticsApi` registered in the app instead of this workaround. Also a bit worried about catching the hook error as we've about to allow APIs to be provided after the initial load.

iamEAP replied:
>  _We're about to allow APIs to be provided after the initial load._

👀  Interesting!  What's the use-case?

I ended up going down this route because, I'd really like the install for analytics plugins to be as simple as possible (e.g. "just add `export { someAnalyticsModule } from '@backstage/plugin-analytics-module-some'` to `app/src/plugins.ts` and you're off to the races!"

But this style of installation is not possible right now with a default, no-op implementation registered in the app.  ...Will this _post-load_ API providing allow for this use-case?

Rugvip replied:
It's for a more SaaS oriented Backstage deployment where you can load in plugins more lazily.

What do you think about this installation method? It's a very similar amount of code but also allows more easier customization through code if needed: https://github.com/backstage/backstage/blob/58633d0762995c160743d16d9ebac4dd6bc04bc2/.changeset/brave-shoes-push.md

iamEAP replied:
It's more than a one-liner, but it's fine.

Rugvip commented on packages/core-plugin-api/src/apis/definitions/AnalyticsApi.ts:
```diff
@@ -0,0 +1,120 @@
+   * - hover
+   * - scroll
+   */
+  verb: string;
```
We're thinking that we might as well just call this `action` instead? 😅

Some other options could be `event` or `type`

iamEAP replied:
Good call-out.  My goal with the naming here was to really discourage people from encoding contextual levels of detail into this one bit of information.  It's really tempting to add extra levels of detail here like `search_result_click`, but in order to be able to uniformly analyze events at various levels of detail, those levels of detail need to be broken out and captured as separate fields (e.g. `component = 'search result' && verb = 'click'`).

Rugvip commented on packages/core-plugin-api/src/apis/definitions/AnalyticsApi.ts:
```diff
@@ -0,0 +1,120 @@
+   * - The value that was filtered by
+   * - The text that was searched for
+   */
+  noun: string;
```
Same as above, perhaps `target`, `object`, `subject`?

Rugvip commented on packages/core-plugin-api/src/apis/definitions/AnalyticsApi.ts:
```diff
@@ -0,0 +1,120 @@
+/**
+ * An analytics event combined with domain attributes.
+ */
+export type DomainDecoratedAnalyticsEvent = AnalyticsEvent & {
```
Same thought as a bit earlier here. Not seeing any usage of `AnalyticsEvent` by itself, perhaps we could just merge `domain` into it and get rid of the `DomainDecoratedAnalyticsEvent`?

Rugvip commented on packages/core-plugin-api/src/apis/definitions/AnalyticsApi.ts:
```diff
@@ -0,0 +1,120 @@
+   * could include information about the route, plugin, or component in which
+   * an event was captured.
+   */
+  domain: AnalyticsDomainValue;
```
Wondering a bit about the existence of both `context` and `domain`, how come they're both needed? I'm thinking you could for example move the known domain fields into the context. Is the purpose of the context to be things that are configured more locally?

iamEAP replied:
Yeah, your intuition is correct.  As someone writing an implementation of this API (potentially even an App Integrator), there's a general need to be able to distinguish between 1) information that is narrowly scoped to the event itself (`context`, in the current implementation) , 2) information pertaining to where the event took place (`domain`, in the current implementation), and 3) information about the person who triggered the event (not yet implemented).  Various analytics systems may treat/capture this information differently.

Rugvip replied:
Id did think it was the other way around to begin with tbh, as in `context` being what you call `domain`, given that it's the context within which the event happens 😅. Do you think it would make sense to flip them around and then find an alternate name for the local one, perhaps just `labels` or smth?

iamEAP replied:
Makes sense.  I went with `attributes`.  🤷‍♂️

Rugvip commented on packages/core-plugin-api/src/apis/definitions/AnalyticsApi.ts:
```diff
@@ -0,0 +1,120 @@
+    verb: string,
+    noun: string,
+    value?: number,
+    context?: AnalyticsEventContext,
```
Perhaps the last two could be merged into an options object as they're not used as often?

iamEAP replied:
Yeah, I think that's probably a good idea.

Rugvip commented on plugins/analytics-module-ga/src/plugin.ts:
```diff
@@ -0,0 +1,33 @@
+export const analyticsModuleGA = createPlugin({
+  id: 'analytics-provider-ga',
+  apis: [
+    createApiFactory({
```
Not sure if plugin is the best way to ship this tbh, but I think we can look at some of the other stuff first and perhaps that can direct this

Rugvip reviewed (commented):
We finally got around to do a mob review of this, really sorry about the delay 😅

Overall looks good 👍, ofc a bunch of comments to iterate on :p

iamEAP commented on packages/core-app-api/config.d.ts:
```diff
@@ -65,6 +65,18 @@ export interface Config {
+       * The provider used to collect instrumented events. Further
+       * configuration values depend on the provider specified here.
+       */
+      provider: string;
```
Open question from me, actually: this introduces the convention (`app.analytics.provider` + `app.analytics.{provider}.*`).  Is there any benefit of defining this in `core-app-api`, or is it fine to leave this up to `analytics-module-{provider}` plugins to define themselves?

Rugvip replied:
Nop I think it's best to keep any provider config schema in the modules, although it's best to establish a compatible pattern between them.

iamEAP commented on packages/core-app-api/src/routing/RouteTracker.tsx:
```diff
@@ -0,0 +1,121 @@
+  useEffect(() => {
+    analytics.captureEvent('navigate', `${pathname}${search}${hash}`);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [pathname, search, hash]);
```
If I recall, I was having issues with capturing duplicate navigation events when I had `analytics` listed as a dep, but lemme play around with it some more.

Rugvip replied:
Realizing that's what we discovered later on with the analytics instance not being stable, which it's best if it is imo

iamEAP commented on packages/core-plugin-api/src/analytics/AnalyticsDomain.tsx:
```diff
@@ -0,0 +1,85 @@
+ * @param Component - Component to be wrapped with analytics domain attributes.
+ * @param domain - Analytics domain key/value pairs.
+ */
+export function withAnalyticsDomain<P>(
```
Good thoughts!  This ties in pretty closely to that design goal I mentioned earlier where "building Backstage plugins the right way" should get you all the `<AnalyticsDomain />` metadata an app integrator would want for free.  ...And then I realized that not everything actually gets exported from a plugin as a component/react extension (e.g. the `<Link />` component in core components).

I would love to avoid this HOC.  But on the other hand, I'm not sure opening up `useAnalytics()` to domain overrides is the best approach either.  As an example:

```jsx
export const MyThing = ({ myValue }) => {
  const analytics = useAnalytics({
    domain: {
      componentName: 'MyThing',
    },
  });
  const handleClick = () => analytics.captureEvent('click', myValue);
  return <Thing value={myValue} onClick={handleClick} />
};
```

In the above, there's less of a clear separation between information specific to the click event, and metadata about where the click event took place (easy to imagine a plugin developer being confused about what information to put where and why).  Also, if any events are instrumented within `<Thing />`, they'll have no association to `MyThing`, even though they happen within that context.

Another possibility that comes to mind is: avoid this HOC, avoid extending `useAnalytics()`, but fix extensions so that `componentName` works / is meaningful?  Maybe it doesn't even make sense to always force a `domain.componentName = Link` on every link click event, when having the context of the component above it is more helpful anyway.

Rugvip replied:
Yeah could be better to have it in the `captureEvent` options as well.

+1 to trying to avoid it altogether though if possible. `componentName` is oddly implementation specific too imo

iamEAP replied:
The HOC is dead!

iamEAP commented:
Okay!  Thank you thank you for the mob review @Rugvip and co.  Before I dive back in on another iteration, I responded to a few things and asked a few questions.  Please have a read through / respond back, and then I can take another stab.

Rugvip commented:
@iamEAP Did another round 🇮🇹

iamEAP commented:
Alright.  Collected the feedback above into a task list for myself, which I will be running through today!  Follow along at home.  (Or not)

- [ ] Little improvements
  - [ ] Go ahead and remove core-app-api/config.d.ts changes
  - [ ] Log a warning when there's no useAnalytics API
- [ ] Simplify the Analytics API Types
  - [ ] Get rid of specialised domain types.
  - [ ] Merge `DomainDecoratedAnalyticsEvent` into `AnalyticsEvent`.
  - [ ] Consider renaming verb/noun to action/subject (possibly look at writing a lint rule for the `action` part)
  - [ ] Rename `AnalyticsDomain` -> `AnalyticsContext` (because that's what it is)
  - [ ] Rename the event `context` to something else (labels? extra? metadata?)
  - [ ] Update `captureEvent()` so that `value`/`context` are on an options object of some kind.
- [ ] Stabilize the `useAnalytics()` tracker response
  - [ ] Add `analytics` back as dependency in RouteTracker
- [ ] `componentName`/`extensionName`
  - [ ] Rename `componentName` to e.g. `extensionName` or just `extension`
  - [ ] Try and find a way to not have to supply the name at extension declaration-time.
  - [ ] Get rid of the `withAnalyticsDomain` HOC (let it inherit from extension context)
- [ ] Refactor `getExtensionDomain()`
  - [ ] Add plugin metadata to routeObjectCollector
  - [ ] Pull this info from routeRefs
- [ ] Improve installation method of GA Plugin
  - [ ] Instead of adding to `plugins.ts`, something similar to [ScmAuth](https://github.com/backstage/backstage/blob/58633d0762995c160743d16d9ebac4dd6bc04bc2/.changeset/brave-shoes-push.md)

iamEAP commented:
@Rugvip (and maybe also @benjdlambert): So the fundamental problem with getting access to the name of an extension (specifically a component extension) is that the lazy loading setup breaks the normal ways of determining a react component's name / display name.  We could do something like this:

```diff
let Component: T;
if ('lazy' in options.component) {
  const lazyLoader = options.component.lazy;
  Component = lazy(() =>
    lazyLoader().then(component => ({ default: component })),
  ) as unknown as T;
} else {
  Component = options.component.sync;
}
const componentName =
+  lazyLoader.name ||
  options.name ||
  (Component as { displayName?: string }).displayName ||
  Component.name ||
  'Component';
```

...But then it just switches from extension component names always being `Component` to being `lazy` unless we also make changes everywhere like this:

```diff
export const SomeExtension = somePlugin.provide(
  createComponentExtension({
    component: {
-      lazy: () =>
-        import('./components/SomeExtension').then(m => m.SomeExtension),
+      lazy: function SomeExtension() {
+        return import('./components/SomeExtension').then(m => m.SomeExtension);
+      },
    },
  }),
);
```

Which feels less clear / more magical than just having a name key.  ...We'd want to / have to update all of the extensions in either case.

Whaddya think?

Rugvip commented:
@iamEAP Yeah I don't even think that's the name we would want in the end either right? The best would actually be name of the export, as in `export const SomeExtension =`. But I believe JS runtime says nooooooooooooo...

I think `name` is good to have for other reasons than the analytics API as well, stuff like introspection for docs or tests and whatnot. The only thing it's unfortunate that it has to be repeated a couple of times x)

iamEAP commented on .changeset/analytics-old-skipping-record.md:
```diff
@@ -0,0 +1,23 @@
+---
+'@backstage/core-plugin-api': minor
```
I've occasionally been using `minor` to indicate "big new thing, who dis," but you're right, there's no break.  Updated.

iamEAP commented:
Alright!  Have another look when you're free.  :eyes: on the extension / route tracking in particular would be appreciated.

Some follow-ups once this is merged:
- A nice fast-follow (better to not do it here in what is already way too big a PR) would be to run a codemod on extensions to add the name key based on the exported variable.  (Note: I updated the template used in `backstage-cli create-plugin` to already include it going forward.)
- Create issues (label `help wanted`) for instrumenting some of the primary actions in the core plugins.
- Documentation, documentation, documentation (API reference and also more explanatory).

iamEAP commented:
Alright!  Have another look when you're free, @Rugvip .  :eyes: on the extension / route tracking in particular would be appreciated.

Some follow-ups once this is merged:
- A nice fast-follow (better to not do it here in what is already way too big a PR) would be to run a codemod on extensions to add the name key based on the exported variable.  (Note: I updated the template used in `backstage-cli create-plugin` to already include it going forward.)
- Create issues (label `help wanted`) for instrumenting some of the primary actions in the core plugins.
- Documentation, documentation, documentation (API reference and also more explanatory).

camilaibs commented on packages/core-app-api/src/routing/RouteTracker.tsx:
```diff
@@ -0,0 +1,109 @@
+): CommonAnalyticsContext | {} => {
+  try {
+    const matches = matchRoutes(routes, { pathname });
+    const routeObject = matches
+      ?.filter(
+        match => (match?.route as BackstageRouteObject).routeRefs?.size > 0,
+      )
+      .pop()?.route as BackstageRouteObject;
```
These lines are not clear to me on first reading, can you explain? Maybe add the explanation in a code comment or variable name?

iamEAP replied:
Updated with some docs (and a little fix).  Thanks!

camilaibs commented on packages/core-app-api/src/app/App.test.tsx:
```diff
@@ -322,6 +338,68 @@ describe('Integration Test', () => {
   });
 
+  it('should track route changes via analytics api', async () => {
+    const mockAnalyticsApi = new MockAnalyticsApi();
+    const apis = [createApiFactory(analyticsApiRef, mockAnalyticsApi)];
+    const app = new PrivateAppImpl({
+      apis,
+      defaultApis: [],
+      themes: [
+        {
+          id: 'light',
+          title: 'Light Theme',
+          variant: 'light',
+          theme: lightTheme,
+        },
+      ],
+      icons: defaultAppIcons,
+      plugins: [],
+      components,
+      bindRoutes: ({ bind }) => {
+        bind(plugin1.externalRoutes, {
+          extRouteRef1: plugin1RouteRef,
+          extRouteRef2: plugin2RouteRef,
+        });
+      },
+    });
+
+    const Provider = app.getProvider();
+    const Router = app.getRouter();
```
Do you think providing a wrapper would reduce the boilerplate?
For example:

**file.test.tsx**
```jsx
it('should track route changes via analytics api', () => {
  const { analyticsApi } = renderWithAnalyticsApi(
     <Routes>
        <Route path="/" element={<NavigateComponent />} />
        <Route path="/foo" element={<HiddenComponent path="/foo" />} />
      </Routes>
  );
  expect(analyticsApi).toHaveBeenCapturedEvent({ ... }); // custom Jest Matcher
});
```

iamEAP replied:
This would definitely make this test way clearer, but I'm not sure it makes sense to break it out for just this one test case.  Perhaps if I studied the rest of the test cases in this suite, I could extract a generic `renderInApp()` where one could provide arbitrary...  API implementations?  Plugins?  Componnets?  Bound routes?  Etc.  Then it might make sense.

Arguably the `renderInTestApp()` function in `test-utils` should be able to do this?  ...But that might be a task unto itself.

camilaibs commented on plugins/analytics-module-ga/README.md:
```diff
@@ -0,0 +1,144 @@
+
+export const apis: AnyApiFactory[] = [
+  // Instantiate and register the GA Analytics API Implementation.
+  createApiFactory({
+    api: analyticsApiRef,
+    deps: { configApi: configApiRef },
+    factory: ({ configApi }) => GoogleAnalytics.fromConfig(configApi),
+  }),
```
Good developer experience on this API, congratulations on the design 👏🏻

camilaibs commented on plugins/analytics-module-ga/README.md:
```diff
@@ -0,0 +1,144 @@
+
+1. Install this plugin in your Backstage app:
+   `cd packages/app && yarn add @backstage/plugin-analytics-module-ga`
+2. Add the API implementation to your App and configure the plugin (see below).
```
Dividing it into three steps can make the getting started more logical (fewer back and forth between code and configuration). Do you agree?
1. Installation
2. Configuration
3. Usage example

iamEAP replied:
I think I get what you're going for.  I updated things a bit accordingly.

camilaibs commented on plugins/analytics-module-ga/README.md:
```diff
@@ -0,0 +1,144 @@
+  analytics:
+    provider: ga
+    ga:
+      trackingId: UA-0000000-0
```
Is it possible to have different tracking ids, for example, one for production and the other for development?

camilaibs replied:
Example:
https://github.com/react-ga/react-ga#reactgainitializegatrackingid-options

![image](https://user-images.githubusercontent.com/6290749/134520697-49392cdf-3795-4585-968d-4db0c434b498.png)

iamEAP replied:
I mentioned elsewhere in this PR, but I recommend we start with the simple, single tracking ID config, and if someone needs multiple trackers, we can add that in later.

camilaibs commented on plugins/analytics-module-ga/README.md:
```diff
@@ -0,0 +1,144 @@
+### Plugin Analytics
+
+In order to be able to analyze usage of your Backstage instance _by plugin_, we
+strongly recommend configuring at least one custom dimension to capture Plugin
```
It might be valuable to promote discoverability since we strongly recommend this usage by having a brief of a custom dimension or, at least, a link to the documentation for those unfamiliar with this concept.

camilaibs commented on plugins/analytics-module-ga/README.md:
````diff
@@ -0,0 +1,144 @@
+      trackingId: UA-0000000-0
+```
+
+### Plugin Analytics
````
```suggestion
### More options
```

camilaibs commented on plugins/analytics-module-ga/README.md:
```diff
@@ -0,0 +1,144 @@
+app:
+  analytics:
+    ga:
+      testMode: true # Prevents data being sent to GA
+      debug: true # Logs analytics event to the web console
```
Here do you mean to have an `app-config.local.yaml` containing this configuration?
If so, would it be good to mention the configuration split or not?

camilaibs reviewed (commented):
Added more comments, will continue reviewing throughout the day.

camilaibs commented on plugins/analytics-module-ga/src/apis/implementations/AnalyticsApi/GoogleAnalytics.ts:
```diff
@@ -0,0 +1,160 @@
+    context: AnalyticsContextValue,
+    attributes: AnalyticsEventAttributes = {},
+  ) {
+    const dataArray = this.cdmConfig
+      .map(cdm => {
+        const value =
+          cdm.source === 'context' ? context[cdm.key] : attributes[cdm.key];
+
+        // Never pass a non-numeric value on a metric.
+        if (cdm.type === 'metric' && typeof value !== 'number') {
+          return undefined;
+        }
+
+        return value !== undefined
+          ? {
+              [`${cdm.type}${cdm.index}`]: value,
+            }
+          : undefined;
+      })
+      .filter(notEmpty);
+
+    return dataArray.length
+      ? dataArray.reduce((result, cd) => {
+          return Object.assign(result, cd);
+        })
+      : {};
+  }
```
I think it can be done in one iteration, something like this:

```ts
 return this.cdmConfig
  .reduce((data, { source, index, type, key }) => {
    const value = source === 'context' ? context[key] : attributes[key];

    // Never pass a non-numeric value on a metric.
    if (type === 'metric' && typeof value !== 'number') {
      return data;
    }
        
    if (value === undefined) {
      return data;
    }
        
    return { ...data, [`${type}${index}`]: value };
  }, {});
```

freben replied:
... or with a `for` loop and some explicit `if`s etc :) i'm not a big fan of big complex reductions for readability, pretty sure they turn out to be massively slower as well, in most cases

iamEAP replied:
Thanks y'all!  Went with a `forEach()` that just appended keys to an object.  A lot clearer!  And a bit more efficient too, perhaps.

camilaibs commented on plugins/analytics-module-ga/src/apis/implementations/AnalyticsApi/GoogleAnalytics.test.ts:
```diff
@@ -0,0 +1,210 @@
+      expect(api.captureEvent).toBeDefined();
+
+      // Initializes GA with tracking ID.
+      expect(ReactGA.testModeAPI.calls[0]).toEqual([
```
This `ReactGA.testModeAPI.calls` is cool maybe we can do something similar with the `MockAnalyticsApi`

iamEAP replied:
Wanted to keep `MockAnalyticsApi` similar to other mocked APIs in `test-utils`.  Though I think there are probably some improvements (similar to this) we can make to all of the mocked APIs in there to improve the test DX for all things in Backstage.

freben commented on packages/core-plugin-api/src/extensions/extensions.tsx:
```diff
@@ -131,7 +137,20 @@ export function createReactExtension<
+                          .id,
+                      }
+                    : {}),
+                }}
```
Suggest creating this attributes object outside of the jsx, it becomes so large and hard to read in here.

Also a nicety with js: object spread is really resilient against "improper" things to spread, which it just ignores silently. You can therefore for example do

```ts
...(options.name && { extension: options.name })
// vs
...(options.name ? { extension: options.name } : {});
```

but that's highly a matter of taste :)

freben commented on packages/test-utils/src/testUtils/apis/AnalyticsApi/MockAnalyticsApi.ts:
```diff
@@ -0,0 +1,41 @@
+import { AnalyticsApi, AnalyticsEvent } from '@backstage/core-plugin-api';
+
+export class MockAnalyticsApi implements AnalyticsApi {
+  private events: AnalyticsEvent[] = [];
```
So ... is there a convenient way of making things like this be reset by `jest.reset/clearAllMocks()`? Kinda would be in line with expectations for a consumer! (don't worry about it too much, i just started thinking out loud)

iamEAP replied:
If so, we should also apply it to `MockErrorApi`, upon which this is based.  Although I think in actual usage, you end up just instantiating new mocks?

freben commented on plugins/analytics-module-ga/config.d.ts:
```diff
@@ -0,0 +1,88 @@
+ */
+
+export interface Config {
+  app: {
```
so this is inside app, is it reasonable to expect that we'll have entirely orthogonal analytics for the backend? i think it's a yes, but wanted to just put it out there anyway :)

iamEAP replied:
Yep!

freben commented on plugins/analytics-module-ga/package.json:
```diff
@@ -0,0 +1,53 @@
+  "version": "0.1.1",
+  "main": "src/index.ts",
+  "types": "src/index.ts",
+  "license": "Apache-2.0",
```
`"private": false`

freben commented on plugins/analytics-module-ga/src/apis/implementations/AnalyticsApi/GoogleAnalytics.ts:
```diff
@@ -0,0 +1,160 @@
+  static fromConfig(config: Config) {
+    // Get all necessary configuration.
+    const trackingId = config.getString('app.analytics.ga.trackingId');
+    const debug = !!config.getOptionalBoolean('app.analytics.ga.debug');
```
it'd be clearer to do `?? false` on these instead of the double exclamation marks!!

freben reviewed (commented):
Alright, FINALLY got to looking through it all in detail! I think it's excellent and these are kinda nitpicky and primarily optional, so don't be discouraged.

iamEAP commented on packages/core-app-api/src/routing/RouteTracker.tsx:
```diff
@@ -0,0 +1,117 @@
+    // the BackstageRouteObject.
+    const routeObject = matches
+      ?.filter(
+        match => (match?.route as BackstageRouteObject).routeRefs?.size > 0,
```
Ah, neat.  I always end up creating little filter type guards off to the side--didn't think to inline the guard.

In this case, it looks like not the approach to use, though, since react-router's `matchRoutes()` returns type `RouteMatch[] | null`, and there isn't a built-in way to tell RR that `RouteMatch.route` is of type `BackstageRouteObject` rather than just `RouteObject`.

Though I guess one type cast on the `matchRoutes()` return is better than the two we've got going on here.  Will update for a little more clarity!

iamEAP commented on packages/core-plugin-api/src/analytics/types.ts:
```diff
@@ -0,0 +1,49 @@
+  /**
+   * The name of the associated extension.
+   */
+  extension: string;
```
Now that I'm thinking of it, I'm wondering if the `Partial<>` is even necessary.  All the attributes of `CommonAnalyticsContext` will always be set (because of context default values).  So any consumer of that type should be able to rely on them being set.

iamEAP commented on packages/core-plugin-api/src/analytics/useAnalytics.tsx:
```diff
@@ -0,0 +1,44 @@
+
+function useTracker(): AnalyticsTracker {
+  const analyticsApi = useApi(analyticsApiRef);
+  const tracker = useRef(new Tracker(analyticsApi));
```
I guess `useRef()` doesn't have this same instantiation shortcut that `useState()` does, but you can achieve the same [this way](https://reactjs.org/docs/hooks-faq.html#how-to-create-expensive-objects-lazily), which I've done.

iamEAP commented on plugins/analytics-module-ga/src/apis/implementations/AnalyticsApi/GoogleAnalytics.ts:
```diff
@@ -0,0 +1,160 @@
+    this.cdmConfig = cdmConfig;
+
+    // Initialize Google Analytics.
+    ReactGA.initialize(trackingId, { testMode, debug, titleCase: false });
```
Thinking we can extend based on additions to the accepted config in the future.

iamEAP commented on .changeset/analytics-haunts-dismembered-constellations.md:
```diff
@@ -0,0 +1,20 @@
+    "extension": "SomeAssociatedExtension",
+    "pluginId": "plugin-in-which-link-was-clicked",
+    "routeRef": "any-associated-route-ref-id"
+  }
```
Easy win: clarified the routeRef/extension bits.

This made me think, though: an important bit of information (perhaps the more important bit of information) is the link _text_.  If you have a series of events (`navigate`, `click`, `navigate`), then you inherently get the URL that was clicked to (in the form of the subject of the second `navigate` event, assuming it's an internal link).

Given that: I updated the implementation to send the link text by default, and the `to` prop as an attribute of the event.

iamEAP commented on .changeset/analytics-sings-stormy-weather.md:
```diff
@@ -0,0 +1,24 @@
+  "context": {
+    "extension": "App",
+    "pluginId": "id-of-plugin-that-exported-the-route",
+    "routeRef": "associated-route-ref-id"
```
An interesting thought.  Answer related to the above: I think in most realistic situations, this is sort of redundant information (e.g. given a series of `navigate` events, you can infer the `from` values).

iamEAP commented on packages/core-components/src/components/Link/Link.test.tsx:
```diff
@@ -40,6 +42,35 @@ describe('<Link />', () => {
+      ),
+    );
+
+    await act(async () => {
```
Yep!  Turns out you can omit the `act()` if you put the expectations in `waitFor()`.

iamEAP commented on plugins/analytics-module-ga/src/apis/implementations/AnalyticsApi/GoogleAnalytics.ts:
```diff
@@ -0,0 +1,160 @@
+  }: {
+    cdmConfig: CDM[];
+    trackingId: string;
+    testMode: boolean;
+    debug: boolean;
```
I think the `fromConfig()` always ends up passing a boolean, regardless of what you have configured, and the constructor itself is private otherwise.  Kept as-is.  Thanks!

iamEAP commented on plugins/analytics-module-ga/src/apis/implementations/AnalyticsApi/GoogleAnalytics.ts:
```diff
@@ -0,0 +1,160 @@
+} from '@backstage/core-plugin-api';
+import { Config } from '@backstage/config';
+
+type CDM = {
```
Good thinking!

Rugvip commented on packages/core-plugin-api/src/extensions/extensions.tsx:
```diff
@@ -127,11 +133,22 @@ export function createReactExtension<
       const Result: any = (props: any) => {
         const app = useApp();
         const { Progress } = app.getComponents();
+        const mountPoint = data?.['core.mountPoint'] as
```
This is a bit unfortunate and won't work if the mount point isn't defined through this `data` but a separate `attachComponentData` call. Let's not try to fix it now but could you add a TODO?

Rugvip commented on plugins/analytics-module-ga/config.d.ts:
```diff
@@ -0,0 +1,88 @@
+export interface Config {
+  app: {
+    // TODO: Only marked as optional because backstage-cli config:check in the
+    // context of the monorepo is too strict. Ideally, this would be marked as
```
Yeah this ends up just being an unfortunate thing with us wanting to have multiple analytics providers installed in the repo at once 😅 Perhaps that's not completely insane for other projects too though? Could be good to make sure the config schema is structured in a way that allows for multiple providers? With that I mean we should remove the `provider: 'ga'` as it'd conflict and is probably redundant anyway? I'm thinking if `analytics.ga` exists, then it is enabled.

Having looked through this I realize that `app.analytics.provider` is actually unused right? In which case it should be nuked from orbit

Rugvip reviewed (commented):
🎉 👏  I think we're pretty much good to go. There are some things I think would be good to fix before merge but can leave stuff as followup as well.

iamEAP commented on .github/styles/vocab.txt:
```diff
@@ -88,6 +88,7 @@ firehydrant
 FireHydrant
 Firekube
 Fiverr
+ga
```
Apparently we did!  But I just rewrote the GA Module readme to not rely on it instead.

iamEAP commented on packages/core-app-api/src/routing/RouteTracker.tsx:
```diff
@@ -0,0 +1,117 @@
+export const RouteTracker = ({ tree }: { tree: React.ReactNode }) => {
+  const { pathname, search, hash } = useLocation();
+  const { routeObjects } = useMemo(() => {
+    return traverseElementTree({
```
Yeah that makes sense.  I added a todo and will create a follow-up issue.

iamEAP commented on packages/core-plugin-api/src/analytics/AnalyticsContext.tsx:
```diff
@@ -0,0 +1,87 @@
+  children: ReactNode;
+}) => {
+  const parentValues = useAnalyticsContext();
+  const combinedValue = useMemo(
```
Yeah, got rid of the memo.  Can explore the functional approach more in the future, but want to avoid getting into a situation where devs could inadvertently start relying on analytics context and changing behavior based on it.

iamEAP commented on packages/core-plugin-api/src/analytics/AnalyticsContext.tsx:
```diff
@@ -0,0 +1,84 @@
+import React, { createContext, ReactNode, useContext } from 'react';
+import { AnalyticsContextValue } from './types';
+
+const AnalyticsReactContext = createContext<AnalyticsContextValue>({
```
Created a todo for the version bridge-ification.  I wonder if future context changes could warrant separate contexts.  While we are saying this is _the_ analytics context, we _actually_ just mean it's contextual information for analytics, and it happens to take the form of a context.  Future stuff might be more appropriate elsewhere anyway.

Rugvip replied:
I'd say let's rewrite it to be a version bridged context after merge but before the release, in case we want to make changes to this context in the future. It might be that there could be additional contexts too, but we don't wanna lock ourselves out of evolving this one. It also prevents this context from breaking when used across multiple versions of `core-plugin-api`

iamEAP commented on packages/core-plugin-api/src/analytics/useAnalytics.tsx:
```diff
@@ -0,0 +1,54 @@
+  // available. Having no default Analytics API implementation enables simple
+  // provider installation via plugin instantiation.
+  try {
+    return useTracker();
```
A downside of this is that anything that uses `<Link />` under the hood (and possibly other core components in the future) has to be wrapped in a test app in storybook.  Maybe a small price to pay.

iamEAP replied:
Yeaaaah, it's pretty painful.  It kinda makes the `@backstage/core-components` update a breaking change.  It doesn't affect anything in real-world apps, but if you have a plugin that happens to use `<Link />` in one of its components (even transiently, like if you use an empty state that uses the link component), and you have a test that renders that component, your tests will break until you wrap it in `wrapInTestApp()`.

iamEAP commented:
Alright. Whaddya think @Rugvip?

Rugvip commented on packages/core-plugin-api/src/analytics/useAnalytics.tsx:
```diff
@@ -0,0 +1,41 @@
+ */
+export function useAnalytics(): AnalyticsTracker {
+  const trackerRef = useRef<Tracker | null>(null);
+  const analyticsApi = useApi(analyticsApiRef);
```
Regarding the `useTracker` discussion, it's very true that that's a bit of a PITA. Let's go with a mix instead? Keep this and the default implementation, because we don't really want plugins to be overriding API implementations willy-nilly under the hood anyway. Then we can keep the presence of the API optional by using `useApiHolder().get(analyticsApiRef)` instead of `useApi`, which will return `undefined` if there's no implementation.

iamEAP replied:
Hmm @Rugvip...  We still end up breaking things in tests downstream because `useApiHolder` expects a versioned API context.

```
    API context is not available

      22 |   const versionedHolder = useVersionedContext<{ 1: ApiHolder }>('api-context');
      23 |   if (!versionedHolder) {
    > 24 |     throw new Error('API context is not available');
         |           ^
      25 |   }
      26 |
      27 |   const apiHolder = versionedHolder.atVersion(1);
```

Is

iamEAP replied:
Kinda dirty, but I can basically re-implement `useApiHolder()`:

```diff
diff --git a/packages/core-plugin-api/src/analytics/useAnalytics.tsx b/packages/core-plugin-api/src/analytics/useAnalytics.tsx
index c3a1ae3b9c..306eca9822 100644
--- a/packages/core-plugin-api/src/analytics/useAnalytics.tsx
+++ b/packages/core-plugin-api/src/analytics/useAnalytics.tsx
@@ -15,16 +15,28 @@
  */
 
 import { useAnalyticsContext } from './AnalyticsContext';
-import { analyticsApiRef, AnalyticsTracker, useApi } from '../apis';
+import { analyticsApiRef, AnalyticsTracker, AnalyticsApi, ApiHolder } from '../apis';
 import { useRef } from 'react';
 import { Tracker } from './Tracker';
+import { useVersionedContext } from '@backstage/version-bridge';
+
+function useTracker(): AnalyticsApi {
+  // We effectively reimplement useApiHolder in order to avoid hook rule order
+  // issues so that the analytics API is truly optional in any consuming code.
+  const versionedHolder = useVersionedContext<{ 1: ApiHolder }>('api-context');
+  const apiHolder = versionedHolder?.atVersion(1);
+  return apiHolder?.get(analyticsApiRef) || {
+    // If no analyticsApi is defined, provide a no-op tracker.
+    captureEvent: () => {},
+  };
+};
 
 /**
  * Get a pre-configured analytics tracker.
  */
 export function useAnalytics(): AnalyticsTracker {
   const trackerRef = useRef<Tracker | null>(null);
-  const analyticsApi = useApi(analyticsApiRef);
+  const analyticsApi = useTracker();
   const context = useAnalyticsContext();
 
   function getTracker(): Tracker {
```

Rugvip replied:
Ah shiet that's true. Tbh I actually prefer the `try ... catch` at that point, but in a way where we take care to have the same hooks being called regardless of path. My worry with reimplementing the versioned thing is that we would forget to update this site whenever we update the `useApiHolder` implementation

iamEAP replied:
Sounds good.  Check it out now?

I'll book some time with you to mob on the version bridge.  I also have a codemod waiting at the ready to get the extension names in place.

iamEAP replied:
Sounds good.  Check it out now, @Rugvip?

I'll book some time with you to mob on the version bridge.  I also have a codemod waiting at the ready to get the extension names in place.

Rugvip reviewed (commented):
+1 about `useAnalytics` breaking stuff, let's fix and then :shipit:

Rugvip reviewed (approved):
Yep, looks good! Let's :shipit: 😁
